### DISCUSSION TITLE
Dragonfly crash in k8s

### DISCUSSION THREAD

**Discussion by Smolsada** [dragonflydb/dragonfly#1481] [General]
Sometimes I see a dragonfly crash for no apparent reason. Is it something well known?

dragonfly chart version: v1.4.0

In logs:
```
PC: @     0x563beaffca6b  (unknown)  util::fb2::intrusive_ptr_release()
*** SIGSEGV received at time=1687353302 on cpu 1 ***
```

**Comment by royjacobson** [comment]
We've seen some related crashes recently and are working on a fix, but it's hard to be sure if they're the same ones from this error alone... Do you happen to use blocking commands? (`BLPOP` and others)

**Reply by Smolsada** [reply]
So far, i'm only using dragonfly for redlock.  I didn't go into the implementation, but i'm almost sure that lua script and setnx are used under the hood.

**Comment by adiholden** [comment]
Hi @Letfdsada ,
I believe that we did not see this specific crash before. 
What is the machine you are running on? are you running on x86 or Arm? which library you are using for redlock?
Any additional information that can help us to reproduce your crash will be helpful

**Reply by Smolsada** [reply]
Arch is amd64 (Intel Xeon Gold 6338).
[Library](https://github.com/madelson/DistributedLock/tree/master/src/DistributedLock.Redis). It seems like all redis calls are contained [here](https://github.com/madelson/DistributedLock/blob/master/src/DistributedLock.Redis/Primitives/RedisReaderWriterLockPrimitives.cs).

I can also change logging level. But I'm not sure that we will see something new related to error. Also, I have no idea how to reproduce this error.

**Reply by romange** [reply]
So it happens rarely? If you still run dragonfly with this redlock lib, can you please run `redis-cli info all` on this instance? This will help us learn which commands are being used, among other things.

**Reply by Smolsada** [reply]
Yes, once or twice in about two weeks.
Output here
```
# Server
redis_version:6.2.11
dfly_version:df-v1.4.0
redis_mode:standalone
arch_bits:64
multiplexing_api:epoll
tcp_port:6379
uptime_in_seconds:968702
uptime_in_days:11

# Clients
connected_clients:18
client_read_buf_capacity:49664
blocked_clients:0

# Memory
used_memory:356056
used_memory_human:347.7KiB
used_memory_peak:370504
comitted_memory:50577408
used_memory_rss:28123136
used_memory_rss_human:26.82MiB
object_used_memory:48768
table_used_memory:281904
num_buckets:6720
num_entries:360
inline_keys:0
strval_bytes:26864
updateval_amount:0
listpack_blobs:0
listpack_bytes:0
small_string_bytes:48768
pipeline_cache_bytes:126
maxmemory:1048576000
maxmemory_human:1000.00MiB
cache_mode:store
maxmemory_policy:noeviction

# Stats
total_connections_received:193963
total_commands_processed:67086682
instantaneous_ops_per_sec:94
total_pipelined_commands:3782773
total_net_input_bytes:8518477935
total_net_output_bytes:69103762
instantaneous_input_kbps:-1
instantaneous_output_kbps:-1
rejected_connections:-1
expired_keys:165
evicted_keys:0
hard_evictions:0
garbage_checked:0
garbage_collected:0
bump_ups:0
stash_unloaded:0
oom_rejections:0
traverse_ttl_sec:640
delete_ttl_sec:0
keyspace_hits:47547374
keyspace_misses:7565
total_reads_processed:26905543
total_writes_processed:10763071
async_writes_count:0
defrag_attempt_total:0
defrag_realloc_total:0
defrag_task_invocation_total:0

# TIERED
tiered_entries:0
tiered_bytes:0
tiered_reads:0
tiered_writes:0
tiered_reserved:0
tiered_capacity:0
tiered_aborted_write_total:0
tiered_flush_skip_total:0

# PERSISTENCE
last_save:1687353303
last_save_duration_sec:0
last_save_file:

# Replication
role:master
connected_slaves:0
master_replid:b434906c7f5dbb615d21c215e7b084638544f0cf

# Commandstats
unknown_SENTINEL:111
cmd_CLIENT:222
cmd_CLUSTER:111
cmd_CONFIG:333
cmd_DEL:6517
cmd_ECHO:222
cmd_EVAL:896
cmd_EVALSHA:18770406
cmd_EXISTS:9
cmd_GET:18771512
cmd_GETRANGE:1
cmd_INFO:128588
cmd_MEMORY:3
cmd_PEXPIRE:18764407
cmd_PING:629629
cmd_QUIT:206
cmd_SCAN:9
cmd_SCRIPT:896
cmd_SET:10011478
cmd_STRLEN:3
cmd_SUBSCRIBE:111
cmd_TTL:2
cmd_TYPE:1010

# Errorstats
unknown_cmd:111
-NOSCRIPT No matching script. Please use EVAL.:3
Cluster is disabled. Enabled via passing --cluster_mode=emulated|yes:111

# Keyspace
db0:keys=360,expires=360,avg_ttl=-1

# CPU
used_cpu_sys:29334.79460
used_cpu_user:56633.150782
used_cpu_sys_children:0.1239
used_cpu_user_children:0.959
used_cpu_sys_main_thread:28229.18186
used_cpu_user_main_thread:55270.357107

# Cluster
cluster_enabled:0
```

**Reply by romange** [reply]
@Letfdsada Thank you! you mentioned logging. It would help us if you could record the commands by adding `--vmodule=main_service=2` to dragonfly command-line arguments.
The commands will be stored inside `dragonfly.INFO` log or in stderr depending on your pod configuration. The log file contains 
line `Starting dragonfly df-1.4.0` in the beginning.
It could help us if you could run dragonfly for a few hours and aggregate commands in the log file and share the log with us.
Then we could try and replay it locally in the hope we will be able to reproduce the failure.

**Reply by adiholden** [reply]
@Letfdsada 
Can you please run `SCRIPT LIST` on the instance and paste here the output?
Are there more logs from the crash ? can you add them as well?
Another request , could you run dragonfly with flag  multi_exec_squash=false , and see if you still experience the crash. We have entered some optimisation lately for scripts and I want to understand if the problem is coming from this flow.

**Reply by Smolsada** [reply]
@adiholden 

> Can you please run SCRIPT LIST on the instance and paste here the output?

```
root@dragonfly-stable-654fb7d9cc-mclz2:/data# redis-cli
127.0.0.1:6379> SCRIPT LIST
1) 1) "6294299faa74c6d411ac7d0da9416eef4a07d268"
   2) "if redis.call('get', ARGV[1]) == ARGV[2] then return redis.call('pexpire', ARGV[1], ARGV[3]) end return 0"
2) 1) "d0dada9c1dfc9ec23dd124919a35b6e72da5a780"
   2) "if redis.call('get', ARGV[1]) == ARGV[2] then return redis.call('del', ARGV[1]) end return 0"
```

> Are there more logs from the crash ? can you add them as well?

Full log on instance:
```
I20230621 13:14:04.724088     1 init.cc:69] dragonfly running in opt mode.
I20230621 13:14:04.724201     1 dfly_main.cc:618] Starting dragonfly df-v1.4.0-6d4d740d6e2a060cbbbecd987ee438cc6e60de79

I20230621 13:14:04.724359     1 dfly_main.cc:671] Max memory limit is: 1000.00MiB
W20230621 13:14:04.724385     1 dfly_main.cc:428] Kernel is older than 5.10, switching to epoll engine.
I20230621 13:14:04.728536     1 proactor_pool.cc:147] Running 2 io threads
I20230621 13:14:04.738384     1 server_family.cc:161] Data directory is "/data"
I20230621 13:14:04.740552     9 listener_interface.cc:96] sock[8] AcceptServer - listening on port 6379
*** SIGSEGV received at time=1687353302 on cpu 1 ***
PC: @     0x563beaffca6b  (unknown)  util::fb2::intrusive_ptr_release()
I20230621 13:15:03.962841     1 init.cc:69] dragonfly running in opt mode.
I20230621 13:15:03.963016     1 dfly_main.cc:618] Starting dragonfly df-v1.4.0-6d4d740d6e2a060cbbbecd987ee438cc6e60de79
I20230621 13:15:03.963204     1 dfly_main.cc:671] Max memory limit is: 1000.00MiB
W20230621 13:15:03.963236     1 dfly_main.cc:428] Kernel is older than 5.10, switching to epoll engine.
I20230621 13:15:03.967216     1 proactor_pool.cc:147] Running 2 io threads
I20230621 13:15:03.977174     1 server_family.cc:161] Data directory is "/data"
I20230621 13:15:03.978936     9 listener_interface.cc:96] sock[8] AcceptServer - listening on port 6379
E20230701 07:13:13.133087     9 epoll_socket.cc:358] sock[17] Unexpected error 110/Connection timed out 0.0.0.0:0
W20230701 07:13:13.133185     9 dragonfly_connection.cc:531] Socket error system:110 Connection timed out
```

**Reply by Smolsada** [reply]
@romange 
I think i need to set up logging to a file. Because 5000 log lines are written to the terminal per minute. It will be hard to get them all out. I'll try to redirect this week if it needed.

Here is result of the last 5000 rows. I replaced the strings that might be sensitive and did a distinct.
```
[PING] in dbid=0
[EVALSHA,MySha,1,MyKey] in dbid=0
LUA [GET,MyKey] in dbid=0
LUA [PEXPIRE,MyKey] in dbid=0
[INFO,replication] in dbid=0
[SET,MyKey] in dbid=0
```

**Reply by romange** [reply]
The annotated log with the failure stack trace is useful. Please send us more if the failure occurs again

**Comment by Smolsada** [comment]
I am with some news.
After `multi_exec_squash=false` and with new versions (we update chart automatically), crashes are eliminated or at least appears less frequently. Haven't seen one in the last 12 days.

**Reply by romange** [reply]
Can you confirm that with `multi_exec_squash=true` the probability of crashes increases?

**Reply by Smolsada** [reply]
I will turn it on today in staging env and write in two weeks if there are no crashes.

**Reply by romange** [reply]
@dranikpg  FYI

**Reply by Smolsada** [reply]
Surprisingly, i never turned off flag on staging :)
For a month or more, not a single crash happened.
Therefore, it seems that matter is not in the flag.
I'm sorry, it's been a tough week. A little later, i will look at the k8s metrics to understand when the crashes stopped.

**Reply by romange** [reply]
We fixed quite a few bugs :)

**Reply by dranikpg** [reply]
multi_exec_squash if now off by default, isn't it? (besides script squashing is not triggered with this flag)

**Reply by Smolsada** [reply]
I'm sorry for late reply. Stuck at work :(

Unfortunately, I didn’t have a metric that would show restarts due to crashes. Because for old metrics we greatly reduce the accuracy.

So it can be said that something has changed since 1.7 (inclusive, based on the history of this discussion)

![image](https://github.com/dragonflydb/dragonfly/assets/33781732/6e8b9707-6553-492e-afaa-323c0a09f748)

**Comment by romange** [comment]
closing this discussion. if you see any bugs or crashes, please open an issue
